Implement stdout columns and rows

sebastienfilion opened the pull request:
To be able to build rich CLI tools, it is very useful to be able to calculate the terminal's number of columns and rows.
Although we could use tput through `Deno.run`, this approach is a bit slow and convoluted.

I know there was already an attempt to solve this issue but it seems to have died out.

This PR showcases two approaches:

1. Exposing ioctl winsize
```ts
const { columns, rows } await Deno.getConsoleSize(Deno.stdout.rid);
... 
const { innerHeight, innerWidth } await Deno.getConsoleSize(Deno.stdout.rid);
```

~~2. Adding properties to STDOUT like in Node.JS~~
```ts
const { columns, rows } = Deno.stdout; 
```

Note: The PR was only tested on a UNIX system.

Resolves: https://github.com/denoland/deno/issues/3456
References: https://github.com/denoland/deno/pull/3790

caspervonb commented on cli/js/ops/ioctl.ts:
```diff
@@ -0,0 +1,5 @@
+import { sendSync } from "./dispatch_json.ts";
+
+export function winsize(rid: number): [number, number] {
```
Should live in tty ops?

caspervonb commented on cli/ops/ioctl.rs:
```diff
@@ -0,0 +1,61 @@
+  let args: WinsizeArgs = serde_json::from_value(args)?;
+  let rid = args.rid;
+
+  let winsize = unsafe { get_dimensions_any(rid) };
```
#[cfg(unix)]

sebastienfilion commented:
@caspervonb Thank you very much for all your feedback! I was expecting some of your comments but with your guidance, I will be more effective!
I think I have the snippet to support windows, I will try to find a way to test it.

As I mentioned in the description of the PR, I've already implemented in on the global namespace as:
```ts
const [ columns, rows ] = await Deno.winsize(Deno.stdout.rid);
```
So, you think that `getConsoleSize` is a better name; Might be better than `getWindowSize` (as in clib) since it could be confused with a browser window...

sebastienfilion commented:
@caspervonb I made the changes that you requested.
For now, I'm throwing `OpError::not_implemented()` for Windows... I have an old Surface somewhere to test an implementation... But I think this can be another PR...

I'd like your opinion on a couple of things:
  1. I added a documentation block along with my function declaration `declare namespace Deno...`; Can you think of another place where I should be documenting this function?
  1. Would you write any unit test for this function?

sebastienfilion commented on cli/ops/tty.rs:
```diff
@@ -250,3 +252,39 @@ pub fn op_isatty(
+  }
+  #[cfg(unix)]
+  {
+    use crate::ops::ioctl::get_dimensions_any;
```
I was thinking that a new file would be useful if we wanted to implement other ioctl commands and not "pollute" the TTY file with libc stuff.

sebastienfilion commented:
> Missing tests and a few nits.

I asked in a previous comment about what tests you would implement; aside from testing that the method is available on the Deno namespace...
I don't know where to get started if we need to mock a console 😆

caspervonb commented:
> I asked in a previous comment about what tests you would implement;

Think we can at-least assert that the values are >= 0 without breaking out a ptty.

ry commented on cli/ops/tty.rs:
```diff
@@ -250,3 +254,70 @@ pub fn op_isatty(
+#[cfg(unix)]
+extern "C" {
+  fn ioctl(fd: c_int, request: c_ulong, ...) -> c_int;
+}
```
Is this `extern "C"` necessary? Why can't you use the one in libc? 
https://docs.rs/libc/0.2.71/libc/fn.ioctl.html

ry commented on cli/js/lib.deno.unstable.d.ts:
```diff
@@ -43,6 +43,26 @@ declare namespace Deno {
+    innerHeight: number;
+    innerWidth: number;
+    rows: number;
+  };
```
Good this is in unstable - thanks!

ry commented:
Hi @sebastienfilion - nice patch!

I agree with @caspervonb that it still needs some work, but this is close to something we can land.

sebastienfilion commented:
Alright, so I "finished" implementing the Windows counterpart of the feature and reworked the code a little.
Now, I'm having issues with the handle on Windows for the STDOUT -- The handle returned from `cli/ops/tty.rs#get_windows_handle` for `rid` 1 and the handle returned from `winapi::um::processenv::GetStdHandle(winapi::um::winbase::STD_OUTPUT_HANDLE)` are different... The former returns `{ columns: 0, rows: 0 }`, which is incorrect 😄 
I can just omit the `rid` on both UNIX and Windows. But I would appreciate it if anyone was able to find an issue with my previous implementation - which I left in comments for the sake of this conversation.

On another note, by build now fails because of something with `cli/tsc.rs` -- I rebased my branch with deno/master

ry commented on cli/ops/tty.rs:
```diff
@@ -250,3 +251,113 @@ pub fn op_isatty(
+  //     _ => Ok(false),
+  //   },
+  // )
+  // .expect("Could not get console screen buffer info.");
```
Dead code should be removed before landing

sebastienfilion replied:
I left this code in reference to my last comment to show my previous implementation; I will remove it before merging.

ry commented on cli/ops/tty.rs:
```diff
@@ -250,3 +251,113 @@ pub fn op_isatty(
+  use std::mem::MaybeUninit;
+
+  let mut count: u32 = 0;
+  loop {
```
What's the purpose of this loop? I think this is unnecessary. ioctl should be called exactly once.

sebastienfilion replied:
There seem to be a chance that ioctl doesn't return the right answer the first time. Every implementation I found uses a loop, here's the code from Node.JS:
```js
  do
    err = ioctl(uv__stream_fd(tty), TIOCGWINSZ, &ws);
  while (err == -1 && errno == EINTR);
```

caspervonb replied:
Interruption of a system call can happen when it is interrupted by signal handler, but I don't see a reason to special case this one system call when all others are just left alone.

If this is something we care about then automatic system call restarts are probably the way to go.

See the "Interruption of system calls and library functions by signal handlers" section in signal(7): http://man7.org/linux/man-pages/man7/signal.7.html

lucacasonato commented:
> I wonder what this function is called in Python, Node, and Go? Maybe we can figure out a common denominator of names.

@ry
- **Python**: `shutil.get_terminal_size`: https://docs.python.org/3/library/shutil.html#querying-the-size-of-the-output-terminal
- **Node**: `process.stdout.columns` and `process.stdout.rows`: https://nodejs.org/api/tty.html
- **Go**: doesn't have a equivalent function

sebastienfilion commented:
> I would be fine if you left Windows not_implemented!() for now.
@ry 
The issue is that the handle extrapolated from the resource table and and the default defined in the `winapi` crate are different. The retrieved handle is `4294967295` and the default handle is `4294967285`...
I ran this to try to compare the addresses `unsafe { winapi::um::processenv::GetStdHandle(get_windows_handle(&std_file)? as u32) } as u32` 

Since the work was done for Windows to be supported - for most use-cases at least - I added an assertion that the `rid` for the console should always be 1 (STDOUT) on Windows; I think it would be an acceptable concession.
If someone, at some point in time, has a use-case requiring `Deno.consoleSize` to handle arbitrary consoles, we can re-open that can of worm.

caspervonb commented:
> winapi::um::processenv::GetStdHandle(get_windows_handle(&std_file)? as u32) } as u32

That doesn't make much sense, GetStdHandle takes constants not file handles...

Anyway just hold on this for a bit please, pushing some cleanup and fixes to @sebastienfilion's branch later.

ry commented on cli/ops/tty.rs:
```diff
@@ -250,3 +251,80 @@ pub fn op_isatty(
+            if libc::ioctl(fd, libc::TIOCGWINSZ, &mut size as *mut _) != 0 {
+              // TODO(caspervonb) this should just be OpError::from(errno)
+              // but libc does not provide errno.
+              return Err(OpError::other("ioctl error".to_owned()));
```
Can you use last_os_error? https://doc.rust-lang.org/std/io/struct.Error.html#method.last_os_error

sebastienfilion commented:
> Preferable to have the verb in there IMHO as this will throw when called on a non tty rid.

I agree with caspervonb.

ry commented on cli/ops/tty.rs:
```diff
@@ -312,7 +314,7 @@ pub fn op_console_size(
             if libc::ioctl(fd, libc::TIOCGWINSZ, &mut size as *mut _) != 0 {
               // TODO(caspervonb) this should just be OpError::from(errno)
               // but libc does not provide errno.
```
Remove TODO
